src/include/txn_inline.h

```diff
@@ -991,8 +994,22 @@ __wt_txn_read_upd_list_internal(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt,
 
         upd_visible = __wt_txn_upd_visible_type(session, upd);
 
-        if (upd_visible == WT_VISIBLE_TRUE)
+        if (upd_visible == WT_VISIBLE_TRUE) {
+            /*
+             * Special case: We previously found a prepared update, check if the visible update we
+             * found has the same transaction id, if it does it must not be visible as it is part of
+             * the same transaction as the previous prepared updated.
+             *
+             * This indicates that the read, which is configured to ignore prepared updates raced
+             * with the commit of the same prepared transaction. If we were to return this update
+             * then the reader would see a partially committed transaction which is bad.
+             */
+            if (prepare_txnid != WT_TXN_NONE && upd->txnid == prepare_txnid) {
+                WT_STAT_CONN_DATA_INCR(session, txn_read_race_prepare_commit);
+                continue;
```

Comment:
That would be closer to the original behaviour but could have the reader waiting, but not for a very long time.

Reply:
I believe both approaches need to compare the previously prepared transaction id to the selected record txnid either in the `txn_read` or internal functions. 

One more approach along the existing PR approach like do not to enter the `__wt_txn_upd_visible_type` function itself if we find the same prepare update transaction id update.

```
if (prepare_txnid == upd->txnid)
   continue;
```

Reply:
I did consider that and admittedly it would be more performant, to keep the statistic I'll also check if that update is committed and then increment my stat.

Reply:
[6db7690](https://github.com/wiredtiger/wiredtiger/pull/9284/commits/6db7690ac7b7e52188a9432801de0adb2fcfdbff)